### PR TITLE
Cumulus 1183: Add lambda configuration check to kes 

### PR DESCRIPTION
**Pull request by Jkovarik**
**Summary:** Summary of changes

Addresses [CUMULUS-1183](https://bugs.earthdata.nasa.gov/browse/CUMULUS-1183)

- **CUMULUS-1183**
  - Kes overrides will now abort with a warning if a workflow step is configured without a corresponding
    lambda configuration

PR also addresses a bug in the test_kes.js unit test setup.

### CODE REVIEW

File: packages/deployment/test/test_kes.js [hunk 1]
```diff
@@ -5,17 +5,19 @@
...
 const get = require('lodash.get');
+const clonedeep = require('lodash.clonedeep');
 
 const configFixture = require('./fixtures/config.json');
 const aliasFixture = require('./fixtures/aliases.json');
 const UpdatedKes = require('../lib/kes');
 
 test.beforeEach((t) => {
-  t.context.kes = new UpdatedKes(configFixture);
+  const config = clonedeep(configFixture);
```
Comment (Jkovarik) [inline packages/deployment/test/test_kes.js]: Kes mutates it's configuration object -- without this, we were causing havoc, particularly a false positive, in later tests l due to the configuration being modified in tests.

File: packages/deployment/test/test_kes.js [hunk 2]
```diff
@@ -187,7 +188,7 @@ test.serial('getAllLambdaAliases returns an unpaginated list of aliases', async
 test.serial('getRetainedLambdaAliasMetadata returns filtered aliasNames', async (t) => {
   const kes = t.context.kes;
 
-  kes.config.workflowLambdas = aliasFixture.workflowLambdas;
+  kes.config.workflowLambdas = clonedeep(aliasFixture.workflowLambdas);
```
Comment (Jkovarik) [inline packages/deployment/test/test_kes.js]: 🤔  we could probably put this in the beforeEach.

File: packages/deployment/test/test_kes.js [hunk 3]
```diff
@@ -217,7 +218,8 @@ test.serial('getRetainedLambdaAliasNames returns filtered aliasNames '
             + 'on previous version redeployment', async (t) => {
   const kes = t.context.kes;
 
-  kes.config.workflowLambdas = aliasFixture.workflowLambdas;
+  kes.config.workflowLambdas = clonedeep(aliasFixture.workflowLambdas);
+  kes.config.maxNumberOfRetainedLambdas = 2;
```
Comment (Jkovarik) [inline packages/deployment/test/test_kes.js]: This was accidently passing due to previous test setup.    Should be set for both getRetainedLambdaAliasMetadata tests.

File: packages/deployment/test/test_kes.js [hunk 4]
```diff
@@ -242,22 +244,22 @@ test.serial('getRetainedLambdaAliasNames returns filtered aliasNames '
   t.deepEqual(expected, actual);
 });
 
-test.serial('getHumanReadableIdentifier returns a packed ID from the descriptiohn string', (t) => {
+test('getHumanReadableIdentifier returns a packed ID from the descriptiohn string', (t) => {
```
Comment (Jkovarik) [inline packages/deployment/test/test_kes.js]: These tests shouldn't need to be serial with proper setup.

File: .eslintrc.json [hunk 5]
```diff
@@ -111,6 +111,7 @@
     "implicit-arrow-linebreak": "off",
 
     "eslint-comments/no-unused-disable": "warn",
+    "eslint-comments/disable-enable-pair": ["error", {"allowWholeFile": true}],
```
Comment (Jkovarik) [inline .eslintrc.json]: This is intended to allow  disabling a rule for the whole file only.  Alternately I can add the disable flag at the end of the file in testConfigvalidation.js.  This seemed like a prudent lint option, but it's definitely up for debate.

File: packages/deployment/test/test_configValidations.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
+/* eslint-disable no-template-curly-in-string */
+/*eslint-comments/disable-enable-pair": ["error", {"allowWholeFile": true}]*/
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/deployment/test/test_configValidations.js]: No, this needs removed.

File: packages/deployment/lib/configValidators.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
+'use strict';
+
+const get = require('lodash.get');
+
+function validateWorkflowDefinedLambdas(config) {
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/deployment/lib/configValidators.js]: Indeed.  Docstring all the things.

File: packages/deployment/test/test_kes.js [hunk 8]
```diff
@@ -5,17 +5,19 @@
...
 
 test.beforeEach((t) => {
-  t.context.kes = new UpdatedKes(configFixture);
+  const config = clonedeep(configFixture);
+  t.context.kes = new UpdatedKes(config);
 });
 
 function setupKesForLookupLambdaReference(kes, functionName, hashValue) {
-  kes.config.lambdas = { [functionName]: { hash: hashValue } };
+  kes.config.lambdas = Object.assign(kes.config.lambdas, { [functionName]: { hash: hashValue } });
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/deployment/test/test_kes.js]: 👍   nice catch, ty.

File: packages/deployment/lib/configValidators.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+  const lambdaResourceMatch = /\$\{(.*)LambdaFunction\.Arn\}/;
+  const stepFunctions = get(config, 'stepFunctions', {});
+  const stepFunctionValues = Object.values(stepFunctions).map((sf) => Object.values(sf.States));
```
[the remark this answers is not in the document]
Comment (Jkovarik) [reply packages/deployment/lib/configValidators.js]: stepFunctionStates

File: packages/deployment/lib/configValidators.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  let resources = [].concat(...stepFunctionStates).reduce((result, cfg) => {
+    if (cfg.Type === 'Task') {
+      const lambdaArnMatch = cfg.Resource.match(lambdaResourceMatch);
+      if (lambdaArnMatch) {
+        result.push(lambdaArnMatch[1]);
+      }
+    }
+    return result;
+  }, []);
+  resources = [...new Set(resources)];
```
Comment (markdboyd) [inline packages/deployment/lib/configValidators.js]: Oooh, `Set`. I haven't seen it used in the wild before. The purpose is to make `resources` unique, right? If so, could the same thing not be accomplished by checking if the `lambdaArnMatch` is already in `result` on line 17 before adding it?

File: packages/deployment/lib/configValidators.js [hunk 11]
```diff
@@ -0,0 +1,32 @@
+'use strict';
+
+const get = require('lodash.get');
+/**
```
Comment (markdboyd) [inline packages/deployment/lib/configValidators.js]: Does ESlint not complain about extra lines between the imports and the docstring?

File: packages/deployment/lib/configValidators.js (hunk 10 above)
Comment (Jkovarik) [reply packages/deployment/lib/configValidators.js]: and in fact, we could throw the error there while we're at it.

Comment (markdboyd) [reply packages/deployment/lib/configValidators.js]: Will the throw "bubble" out of the anonymous function scope for `reduce`?

Comment (Jkovarik) [reply packages/deployment/lib/configValidators.js]: No, but we can return a structure that can evaluate to do or do not throw error.     That's likely less clean though, if more performant.    Performance not being a significant issue here, that's probably not worth it.

Comment (Jkovarik) [reply packages/deployment/lib/configValidators.js]: In either case, good point : 96b53c5

**Pull request merged**